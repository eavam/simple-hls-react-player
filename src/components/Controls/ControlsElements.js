import styled from 'react-emotion';
import { flexContainer, flexLeft, flexRight } from '../styledComponens';

export const Root = styled('div')`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
`;

export const Title = styled('div')`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  height: 100px;
  display: flex;
  align-items: center;
  padding: 0 2%;
  box-shadow: inset 0px 100px 60px -35px black;
  color: #fff;
  font-size: 30px;
`;

export const BottomBar = styled('div')`
  ${flexContainer};
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  height: 100px;
  align-items: center;
  padding: 0 2%;
  box-shadow: inset 0px -100px 60px -35px black;
`;

export const LeftColumn = styled('div')`
  ${flexLeft};
`;

export const RightColumn = styled('div')`
  ${flexRight};
`;
