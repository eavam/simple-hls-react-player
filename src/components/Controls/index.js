import React from 'react';
import styled from 'styled-components';

import TopControlsContainer from '../../containers/TopControlsContainer';
import BottomControlsContainer from '../../containers/BottomControlsContainer';

const Controls = () => (
  <Root>
    <TopControlsContainer />
    <BottomControlsContainer />
  </Root>
);

const Root = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
`;

export default Controls;
