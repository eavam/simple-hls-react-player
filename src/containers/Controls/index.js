import React, { Component } from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import * as PlayerActions from '../../actions/PlayerActions.js'

import Title from '../../components/Title'
import Fullscreen from '../../components/Fullscreen'
import CurrentTime from '../../components/CurrentTime'
import SoundControl from '../../components/SoundControl'
import PlayPauseControl from '../../components/PlayPauseControl'

import ControlsWrapper from '../../layout/ControlsWrapper'
import BottomControlsWrapper from '../../layout/BottomControlsWrapper'
import TopControlsWrapper from '../../layout/TopControlsWrapper'
import LeftColumn from '../../layout/LeftColumn'
import RightColumn from '../../layout/RightColumn'

export class Controls extends Component {

  constructor(props) {
    super(props)
    this.handleChangeVolume = this.handleChangeVolume.bind(this)
    this.volumeMouseMove = this.volumeMouseMove.bind(this)
  }

  /**
   * Вызывает action на изменения ползунка громкости
   * 
   * @param {event} event - вызываемое событие
   */
  handleChangeVolume(event) {
    const bindVolumeMouseMove = this.volumeMouseMove(event)

    // Вызваем сразу чтобы реагировать на одиночный клик
    bindVolumeMouseMove(event)

    // Изменяет громкость при перемищении ползунка
    window.addEventListener('mousemove', bindVolumeMouseMove)

    // При отпускании кнопки убираем listener
    window.addEventListener('mouseup', () => {
      window.removeEventListener('mousemove', bindVolumeMouseMove)
    })
  }

  /**
   * Вычисляет на какое значение нужно изменить громкость
   * 
   * @param {event} event - вызываемое событие
   */
  volumeMouseMove(event) {
    const { actions, videoEl } = this.props

    const { left: zeroPoint , width: maxPoint } = event.target.getBoundingClientRect()

    return (windowEvent) => {

      // Получаем разницу между текущей позицией клика
      // и минимальным возможной позицией
      let newVolume = windowEvent.clientX - zeroPoint

      // Проверяем чтобы значение небыло выше или ниже допустимых
      if( newVolume >= maxPoint ) newVolume = maxPoint
      if( newVolume <= 0 ) newVolume = 0

      // Получаем процентное значение
      const percentVolume = newVolume / (maxPoint / 100)

      // Получаем десятичное значение
      const volumeElementVal = percentVolume / 100

      videoEl.volume = volumeElementVal
      actions.changeVolume(newVolume)

    }
  }

  render() {

    const {
      volume,
      isPause,
      muted,
      currentTime,
      isFullscreen,
      title,
      actions
    } = this.props

    return (
      <ControlsWrapper>

        <TopControlsWrapper>
          <Title title={title} />
        </TopControlsWrapper>

        <BottomControlsWrapper>

          <LeftColumn>
            <PlayPauseControl isPause={isPause} onClick={actions.togglePause} />
            <SoundControl
              handleChangeVolume={this.handleChangeVolume}
              toggleMute={actions.toggleMute}
              volume={volume}
              muted={muted}
            />
          </LeftColumn>

          <RightColumn>
            <CurrentTime time={currentTime} />
            <Fullscreen isFullscreen={isFullscreen} toggleFullscreen={actions.toggleFullscreen} />
          </RightColumn>

        </BottomControlsWrapper>
      </ControlsWrapper>
    )
  }
}

Controls.propTypes = {
  actions: PropTypes.object,
  volume: PropTypes.string,
  muted: PropTypes.bool,
  currentTime: PropTypes.string,
  isFullscreen: PropTypes.bool,
  title: PropTypes.string,
  handleChangeVolume: PropTypes.func,
  isPause: PropTypes.bool,
  videoEl: PropTypes.object
}

function mapStateToProps(state) {

  const {
    videoEl,
    volume,
    isPause,
    muted,
    currentTime,
    isFullscreen,
    title
  } = state

  return {
    videoEl,
    volume,
    isPause,
    muted,
    currentTime,
    isFullscreen,
    title
  }
}

function mapDispathToProps(dispath) {
  return {
    actions: bindActionCreators(PlayerActions, dispath)
  }
}

export default connect(mapStateToProps, mapDispathToProps)(Controls)