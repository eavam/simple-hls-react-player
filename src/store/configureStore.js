import { createStore, applyMiddleware, compose } from 'redux'
import rootReducer from '../reducers'
import thunk from 'redux-thunk'

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose

const configureStore = initialState => {
  const store = createStore(rootReducer, composeEnhancers(applyMiddleware(thunk)))

  if (module.hot) {
    module.hot.accept('../reducers', () => {
      const nextRootReducer = require('../reducers')
      store.replaceReducer(nextRootReducer)
    })
  }

  return store
}

export default configureStore
