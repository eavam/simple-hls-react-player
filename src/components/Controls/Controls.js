// @flow
import React from 'react';
import withControlsConnect from '../../hocs/withControlsConnect';
import { Root, Title, BottomBar, LeftColumn, RightColumn } from './ControlsElements';
import PlayPauseControl from '../PlayPauseControl';
import SoundControl from '../SoundControl';
import CurrentTime from '../CurrentTime';
import Fullscreen from '../Fullscreen';

type Props = {
  title?: string,
  isFullscreen?: boolean,
  volume?: number,
  muted?: boolean,
  time?: string,
  typeActionPlayer: 'play' | 'pause',
  onClickPlay: ?() => void,
  onClickMute: ?() => void,
  onChangeVolume: ?() => void,
  onClickFullScreen: ?() => void,
};

const Controls = ({
  title,
  isFullscreen,
  typeActionPlayer,
  volume,
  muted,
  time,
  onClickPlay,
  onClickMute,
  onChangeVolume,
  onClickFullScreen,
}: Props) => (
  <Root>
    <Title>{title}</Title>
    <BottomBar>
      <LeftColumn>
        <PlayPauseControl typeActionPlayer={typeActionPlayer} onClick={onClickPlay} />
        <SoundControl
          onMouseDown={onChangeVolume}
          onClick={onClickMute}
          volume={volume}
          muted={muted}
        />
      </LeftColumn>

      <RightColumn>
        <CurrentTime time={time} />
        <Fullscreen isFullscreen={isFullscreen} onClick={onClickFullScreen} />
      </RightColumn>
    </BottomBar>
  </Root>
);

Controls.defaultProps = {
  title: '',
  isFullscreen: false,
  volume: 0,
  muted: false,
  time: '00:00',
};

export default withControlsConnect(Controls);
