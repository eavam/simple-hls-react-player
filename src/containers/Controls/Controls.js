/* import React from 'react';
import styled from 'react-emotion'; */
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { createStructuredSelector } from 'reselect';

import Controls from '../../components/Controls';
import {
  getTitle,
  getTypeActionPlayer,
  getVolume,
  getMuted,
  getTime,
  getFullscreen
} from '../../selectors';
import { handleChangeVolume } from '../../actions/PlayerActions';
import { togglePause, toggleMute, toggleFullscreen } from '../../actionTypes';

const mapStateToProps = createStructuredSelector({
  title: getTitle,
  isFullscreen: getFullscreen,
  typeActionPlayer: getTypeActionPlayer,
  volume: getVolume,
  muted: getMuted,
  time: getTime
});

const mapDispatchToProps = dispatch =>
  bindActionCreators({ togglePause, toggleMute, toggleFullscreen, handleChangeVolume }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Controls);
