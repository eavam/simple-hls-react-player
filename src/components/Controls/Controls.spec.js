import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { render } from 'enzyme';
import Controls from './index';

test('Controls default', () => {
  const mockStore = configureStore();
  const store = mockStore({
    typeActionPlayer: 'play',
  });

  const component = renderer.create(
    <Provider store={store}>
      <Controls
        togglePause={() => {}}
        toggleMute={() => {}}
        handleChangeVolume={() => {}}
        toggleFullscreen={() => {}}
      />
    </Provider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Contorls has title', () => {
  const mockStore = configureStore();
  const store = mockStore({
    typeActionPlayer: 'play',
  });

  const component = (
    <Provider store={store}>
      <Controls title="Test title controls" />
    </Provider>
  );

  expect(render(component).text()).toMatch('Test title controls');
});

test('Controls has time', () => {
  const mockStore = configureStore();
  const store = mockStore({
    time: '10:10',
    typeActionPlayer: 'play',
  });

  const component = (
    <Provider store={store}>
      <Controls />
    </Provider>
  );

  expect(render(component).text()).toMatch('10:10');
});
