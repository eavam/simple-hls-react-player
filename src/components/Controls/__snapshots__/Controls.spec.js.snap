// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls default 1`] = `
<div
  className="css-jy6mft"
>
  <div
    className="css-c6kjrz"
  >
    
  </div>
  <div
    className="css-w74n7p"
  >
    <div
      className="css-1wv0v70"
    >
      <div
        className="css-12grdtp"
        onClick={[Function]}
      >
        <svg
          viewBox="0 0 20.03 25"
        >
          <path
            d="M7 21.52a3.48 3.48 0 0 1-7 0v-18a3.48 3.48 0 1 1 7 0zm0 0M20 21.52a3.48 3.48 0 0 1-7 0v-18a3.48 3.48 0 0 1 7 0zm0 0"
            fill="#fff"
          />
        </svg>
      </div>
      <div
        className="css-1sg2lsz"
      >
        <div
          className="css-1htvj9q"
          onClick={[Function]}
          role="presentation"
        >
          <svg
            viewBox="0 0 17.45 24.61"
          >
            <path
              d="M16.52.17a1.64 1.64 0 0 0-1.73.18L9.52 4.48a3.2 3.2 0 0 0-.9-.13H3.21A3.21 3.21 0 0 0 0 7.56v9.36a3.21 3.21 0 0 0 3.21 3.21h5.41a3.22 3.22 0 0 0 .78-.1l5.39 4.23A1.64 1.64 0 0 0 17.45 23V1.64a1.64 1.64 0 0 0-.93-1.47zm0 0"
              fill="#fff"
            />
          </svg>
        </div>
        <div
          className="css-hvewp1"
          muted={false}
          onMouseDown={[Function]}
        />
      </div>
    </div>
    <div
      className="css-owsd35"
    >
      <div
        className="css-mzstqo"
      >
        00:00
      </div>
      <div
        className="css-1yncnvt"
        onClick={[Function]}
      >
        <svg
          viewBox="0 0 25 25"
        >
          <path
            d="M4.54 20.5h16v-16h-16zM6.83 6.79h11.43v11.42H6.83zm0 0"
            fill="#fff"
          />
          <path
            d="M2.28 18.18H0V25h6.82v-2.28H2.28zm0 0M2.28 2.28h4.54V0H0v6.82h2.28zm0 0M22.72 22.72h-4.54V25H25v-6.82h-2.28zm0 0M18.18 0v2.28h4.54v4.54H25V0zm0 0"
            fill="#fff"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
